pkg/sql/opt/table_meta.go

```diff
@@ -210,9 +211,13 @@ type TableMeta struct {
 	// anns annotates the table metadata with arbitrary data.
 	anns [maxTableAnnIDCount]interface{}
 
-	// notVisibleIndexMap stores information about index invisibility which maps
-	// from index ordinal to index invisibility.
-	notVisibleIndexMap map[cat.IndexOrdinal]bool
+	// notVisibleIndexes stores the ordinals of indexes that have been marked as
+	// not visible. This avoids recomputation and ensure a consistent answer
+	// within a query for indexes with fractional visibility.
+	notVisibleIndexes intsets.Fast
+	// cachedIndexVisibility indicates whether the visibility of given index
+	// ordinal has been cached in notVisibleIndexes.
+	cachedIndexVisibility intsets.Fast
```

Comment:
nit: purely stylistic, but this might be a good spot for a nested struct:

```go
  indexVisibility struct {
    notVisible intsets.Fast
    isCached   intsets.Fast
  }
```